### ISSUE TITLE
Possible Unintended Behaviour for Google Play Authentication Issues.

### ISSUE THREAD

**Issue by Fall3nWarrior** [playgameservices/play-games-plugin-for-unity#2531]
Recently I started integrating the Play Games Plugin for Unity in my Game and ran across an issue where the game would fail to authenticate with Google Play Services. It failed 100% of the time. 

After researching the issue further, checking my code implementation, Google Play Console settings and trying some of the steps that others did to remedy the situation, I discovered the problem. Unfortunately for me, none of the steps I found elsewhere solved the issue as my app signing keys matched up and all other settings were correct in the Google Play Console.

It was only after I compared the LollyGagger codelab [here](https://codelabs.developers.google.com/codelabs/playservices_unity/index.html?index=..%2F..index#0) to the build settings of my game, that I found the problem. In Unity under File > Build Settings > Android Platform, if Development Build is checked then Google Play Services fails to authenticate 100% of the time in Development Builds. Therefore the solution is to uncheck the box.

Here is a screenshot of the setting in question.
![Build Settings](https://user-images.githubusercontent.com/49316390/55640497-b9c7c280-5799-11e9-994b-6ad6f7c86098.jpg)

I am not sure if this is intended or a bug. I just wanted to share my solution so that I may help others who have been stuck and to report a possible unintended behaviour or bug.

**Comment by Noah1911** [comment]
Looks like this is intented as in Unity default template for Gradle signs Development build with debug keystore if you uncheck it, then it will use your release keystore. 
Workaround for this is by letting Unity generate custom gradle template file and copypaste **SIGNCONFIG** from release section to debug one.

**Comment by Teddymac22** [comment]
@Fall3nWarrior Thanks so much for verifying this. I seen where someone mentioned this in a post somewhere on the net, but I wasn't sure if this was the case. I have been struggling with this issue myself as shown in a recent post here.

I am working on 2 projects that use the same game engine and one used Admob and the other did not. After updating GPS to 0.9.60 and Facebook SDK to 7.15.1 the one project without Admob worked correctly with Development Mode unchecked. This version has a Beta build out, but no Alpha or Production.

The project with Admob had all kinds of issues so I have finally just removed it and I am now attempting to get GPS working. Now, with this version, If I leave Development Mode unchecked Facebook nor GPS will even attempt to start. With GPS the callback is never used. 
With Development Mode checked, Facebook will work it seems and GPS shows it failed to authenticate. I have put this version into Alpha with the same results. This version has a Beta and a Production out that are both older than the Alpha. 

So, I am wondering do you need to publish a version to Beta in order the the GPS to start authenticating? I will be testing this soon as it seems like in the past that this fixed issues I had before. It has been many months sense I posted an update so I can't remember what I did before.

**Comment by Teddymac22** [comment]
Quick update. I just tested this with another project I have going that uses the same game engine. GPS and Facebook SDK both work for the app if Development Build is left unchecked. So the solution to uncheck the Development Build box does not apply to all situations. This app has a live Beta version but no Production version. So I wonder if the versions there are uploaded to the store factors into what GPS will allow?

**Comment by Noah1911** [comment]
@Teddymac22 Does the solution I proposed above work for you?

Steps to make it happen are the following:
1. Open Publishing Settings in Player Settings (Unity)
2. Check Custom Gradle Template
3. Let Unity generate mainTemplate.gradle file
4. Open the generated file in some editor (ex Notepad++) and modify android -> buildTypes section as following:
```
    buildTypes {
        debug {
            minifyEnabled **MINIFY_DEBUG**
            useProguard **PROGUARD_DEBUG**
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-unity.txt'**USER_PROGUARD****SIGNCONFIG** /* <- this is the key to sign dev build too */
            jniDebuggable true
        }
        release {
            minifyEnabled **MINIFY_RELEASE**
            useProguard **PROGUARD_RELEASE**
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-unity.txt'**USER_PROGUARD****SIGNCONFIG**
        }
    }**PACKAGING_OPTIONS****SPLITS**
```
6. Don't forget to remove Unity comment on first line!

**Comment by Teddymac22** [comment]
@Noah1911 
Oh my goodness, that worked! I am unfamiliar with Android Studio, Gradle, and the like, so I didn't understand what you originally suggested. This also allows for Facebook SDK to function and I bet Admob as well in one of my previous setups. 

Ok, hear is a question...
I have two projects using the same engine, with Project A if I uncheck Development Build, neither GPS or  Facebook functions. But in Project B if I leave Development Build unchecked it works just fine. Project A has a live Production going, but Project B is in Open Beta.

GPS nor Facebook works on Project A when I download from the Play Store the internal or Alpha build. But it works now with a direct build to device and Development Build checked (thanks to Noah1911 solution).

I checked Project A's settings in the console and testing is turned on for Alpha and Internal. Perhaps it takes time for the GPS to kick in? Or do I need to publish it to Beta? The code obviously works as it works now in Development Mode so there has to be some other reason. I do not use the Google Signing feature on either project.

**Comment by Noah1911** [comment]
@Teddymac22 Can you post adb's logcat log?

**Comment by Teddymac22** [comment]
Here are a couple of logcats. One with GPS working in Development mode and and the one not working(logcatMissingGPS) was downloaded from the Google Play Store as a Alpha build. Noting works in the none working one, not even purchases and it has a line that says...
E/Unity   (29704): AndroidJavaException: java.lang.ClassNotFoundException: com.unity.purchasing.googleplay.GooglePlayPurchasing

E/Unity   (29704): java.lang.ClassNotFoundException: com.unity.purchasing.googleplay.GooglePlayPurchasing

So, why are the needed classes not being used?

[logcatWihtGPS.txt](https://github.com/playgameservices/play-games-plugin-for-unity/files/3054649/logcatWihtGPS.txt)
[logcatMissingGPS.txt](https://github.com/playgameservices/play-games-plugin-for-unity/files/3054680/logcatMissingGPS.txt)

**Comment by Fall3nWarrior** [comment]
@Noah1911 Thank you for your explanation and the step by step tutorial. It worked perfectly for me and cleared up any confusion!

**Comment by Fall3nWarrior** [comment]
@Teddymac22 Your welcome for verifying this for you. Unfortunately, I don't think I can help you with your issue as I am new to all of this myself. But I'm sure that @Noah1911 or someone else can help. Good luck!

**Comment by vaibhavgpt-google** [comment]
Closing this issue as part of a cleanup of older tickets. Google Play Games v1 is now deprecated. Please upgrade to version v2.1.0 or newer: https://github.com/playgameservices/play-games-plugin-for-unity/releases/tag/v2.1.0. If this issue persists on the new v2 architecture, please open a new ticket.
